### DISCUSSION TITLE
Checkbox x-model not updated...

### DISCUSSION THREAD

**Discussion by petasis** [alpinejs/alpine#3712] [7. Bugs (archived)] [answered]
Hi all,

I am new to Alpine, so perhaps I am doing something wrong.

I have a table column, that has a set of checkboxes. In the header of the table, I have a checkbox, that is supposed to check/uncheck all of them.

The html of the table header checkbox is:
```
<th x-data="{ toggleSelection(event) {
               select_all = !select_all;
               let checkboxes = document.querySelectorAll('#table-container-Activity input[name^=&quot;selected&quot;]');
               [...checkboxes].map((el) => {
                 el.checked = select_all;
                 $store.dashboardInteractiveFiltering.addPK('Activity', el.value, el.checked);
               });
               } }" @click="toggleSelection()" style="cursor: pointer;">
            
                  <input type="checkbox" class="form-check-input" x-model="select_all">
              </th>
```
Each table row checkbox is:
```
<tr class="even" x-data="{ isSelected: false }" :class="{ 'highlight-me': isSelected }">
        
  <td><input type="checkbox" name="selected" value="6" class="form-check-input" x-model="isSelected" x-text="isSelected" id="selected-Activity" @click="$store.dashboardInteractiveFiltering.addPK(&quot;Activity&quot;, $el.value, $el.checked);"></td>
</tr>
```
If I click on the row checkbox, x-model's isSelected is updated with true/false (its visible).

However, if I check the header checkbox, and the toggleSelection() is called (which sets checked directly), isSelected does not change.

What am I doing wrong?

**Comment by ekwoka** [comment, marked as answer]
You should be modeling all the checkboxes to state that selectAll can directly modify. So that it can really be more like

```
<table x-data="{
  checkboxes: [{ isChecked: true }, { isChecked: false }],
  toggleAll({ currentTarget }) {
      this.checkboxes.forEach(cb => cb.isChecked = currentTarget.checked
  },
}">
  <th>
    <input type="checkbox" @change="toggleAll" :checked="checkboxes.every(cb => cb.isChecked)" :inditerminate="checkboxes.some(cb => cb.isChecked)">
    </th>
     <th>
    <input.type="checkbox" x-model="checkboxes[0]" >
     </th>
     
</th>
</table>
```

Or whatever, you can use an object instead of an array if you want. I don't know what your data is to know the exact best implementation.

But basically, x-model listens on change events, checkboxes do not emit change events when programmatically changed, only when changed in the UI.

**Reply by petasis** [reply]
So, I have to modify the x-model variable for this to work?

There is not way for it to work, if I modify $el.checked?

**Reply by petasis** [reply]
Is there a way to get all checkboxes, and their x-model, without hardcoding them as the checkboxes array?

**Reply by petasis** [reply]
Can I start with checkboxes as an empty array, and each checkbox to append its value to it? Can x-init be used for it?

**Reply by ekwoka** [reply]
You can dispatch change events on all of them.

You can use x-init. model can handle modelling them all to one array, but it won't work well for setting them all true, unless you know all the values.

You can just have the values be an array, and model the checkbox to the array will add and remove it's value from the array.
